File: eval_harness/utils/telemetry/tracing/codex_rollout.py
#!/usr/bin/env python3
"""
Codex rollout -> normalized agent trace (offline, no OTel, no live hook).

The `parse_rollout` state machine is vendored faithfully from the user's
codex_bt_hook.py (which mirrors Langfuse's codex-observability-plugin). The
difference here: instead of running as a live Codex `Stop` hook, this runs at
COLLECTION TIME — it finds the rollout JSONL(s) Codex wrote during a run, parses
them into the normalized turns->steps->tool_calls shape, and writes one JSON
file into the run bundle. That avoids the leftover background process the live
cc-trace/hook approach left behind.

Codex `exec` writes rollouts under $CODEX_HOME/sessions/**/rollout-*.jsonl
(default ~/.codex/sessions). We select rollouts modified at/after --since-mtime
(the orchestrator records the run's start time) so we only pick up THIS run.

Optional: with --braintrust and BRAINTRUST_API_KEY set, also push the spans
(reusing eval_harness/utils/telemetry/tracing/bt_emit.py), tagged with --run-id.

Test the parser standalone:  python codex_rollout.py --parse-only <rollout.jsonl>
"""

import argparse
import glob
import json
import os
import pathlib
import sys


# ---------- rollout parsing (faithful to the Langfuse plugin state machine) ----------

def _extract_text(content) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        out = []
        for part in content:
            if isinstance(part, dict):
                out.append(part.get("text") or part.get("output_text") or part.get("input_text") or "")
            elif isinstance(part, str):
                out.append(part)
        return "".join(out)
    return ""


def _extract_reasoning(payload) -> str:
    c = payload.get("content")
    if c:
        return _extract_text(c)
    summary = payload.get("summary")
    if isinstance(summary, list):
        return " ".join(_extract_text(s) for s in summary)
    return ""


def _is_wrapper(text: str) -> bool:
    t = (text or "").lstrip()
    return t.startswith("<environment_context>") or t.startswith("<user_instructions>")


def _maybe_json(s):
    if isinstance(s, (dict, list)):
        return s
    if isinstance(s, str):
        try:
            return json.loads(s)
        except Exception:
            return s
    return s


def _usage(info):
    if not isinstance(info, dict):
        return {}
    out = {}
    if info.get("input_tokens") is not None:
        out["prompt_tokens"] = info["input_tokens"]
    if info.get("output_tokens") is not None:
        out["completion_tokens"] = info["output_tokens"]
    if info.get("total_tokens") is not None:
        out["tokens"] = info["total_tokens"]
    if info.get("cached_input_tokens") is not None:
        out["cache_read_input_tokens"] = info["cached_input_tokens"]
    return out


def _tool_error(payload):
    return (payload.get("error") or payload.get("codex_error_info")
            or payload.get("stderr") or payload.get("aggregated_output")
            or (f"Exit code: {payload['exit_code']}" if payload.get("exit_code") else "failed"))


def parse_rollout(path):
    records = []
    with open(path) as f:
        for ln in f:
            ln = ln.strip()
            if not ln:
                continue
            try:
                records.append(json.loads(ln))
            except Exception:
                continue  # skip malformed line, don't abort

    session_meta, turns = {}, []
    cur = None        # current MutableTurn
    cur_step = None   # current model step
    tools_by_id = {}
    pending_tool_outputs = {}
    turn_index = 0

    def finish_step():
        nonlocal cur_step
        if cur is not None and cur_step is not None:
            cur["steps"].append(cur_step)
            cur_step = None

    def finish_turn(completed, aborted):
        nonlocal cur, tools_by_id, pending_tool_outputs
        if cur is None:
            return
        finish_step()
        cur["completed"], cur["aborted"] = completed, aborted
        if not cur.get("final_output"):
            for st in reversed(cur["steps"]):
                if st.get("text"):
                    cur["final_output"] = st["text"]
                    break
        if not cur.get("user_input"):
            cur["user_input"] = cur.get("user_input_fallback")
        turns.append(cur)
        cur, tools_by_id, pending_tool_outputs = None, {}, {}

    def tool_output(payload):
        return (
            payload.get("aggregated_output")
            or payload.get("stdout")
            or payload.get("result")
            or payload.get("output")
            or {
                k: payload.get(k)
                for k in ("status", "query", "action", "call_id")
                if payload.get(k) is not None
            }
        )

    def attach_pending(tc):
        call_id = tc.get("call_id")
        if not call_id or call_id not in pending_tool_outputs:
            return
        pending = pending_tool_outputs.pop(call_id)
        if pending.get("error"):
            tc["error"] = pending["error"]
        if pending.get("output") is not None:
            tc["output"] = pending["output"]

    for rec in records:
        typ = rec.get("type")
        payload = rec.get("payload") or {}

        if typ == "session_meta":
            session_meta = payload
        elif typ == "turn_context" and cur is not None:
            cur["model"] = payload.get("model") or cur.get("model")
            cur["invocation_params"] = payload
        elif typ == "response_item":
            if cur is None:
                continue
            if cur_step is None:
                cur_step = {"text": "", "reasoning": "", "tool_calls": []}
            pt = payload.get("type")
            if pt == "message":
                role = payload.get("role")
                text = _extract_text(payload.get("content"))
                if role == "assistant":
                    cur_step["text"] += text
                elif role in ("user", "developer") and not _is_wrapper(text):
                    cur.setdefault("user_input_fallback", text)
            elif pt == "reasoning":
                cur_step["reasoning"] += _extract_reasoning(payload)
            elif pt in ("function_call", "custom_tool_call"):
                tc = {
                    "call_id": payload.get("call_id"),
                    "name": payload.get("name"),
                    "args": _maybe_json(payload.get("arguments") if payload.get("arguments") is not None
                                        else payload.get("input")),
                }
                cur_step["tool_calls"].append(tc)
                if tc["call_id"]:
                    tools_by_id[tc["call_id"]] = tc
                    attach_pending(tc)
            elif pt == "web_search_call":
                action = payload.get("action") or {}
                tc = {
                    "call_id": payload.get("id"),
                    "name": "web_search",
                    "args": {
                        "query": action.get("query"),
                        "queries": action.get("queries"),
                        "action": action,
                    },
                }
                cur_step["tool_calls"].append(tc)
                if tc["call_id"]:
                    tools_by_id[tc["call_id"]] = tc
                    attach_pending(tc)
            elif pt in ("function_call_output", "custom_tool_call_output"):
                tc = tools_by_id.get(payload.get("call_id"))
                if tc is not None:
                    tc["output"] = payload.get("output")
        elif typ == "event_msg":
            pt = payload.get("type")
            if pt == "task_started":
                if cur is not None:
                    finish_turn(completed=False, aborted=False)
                turn_index += 1
                cur = {"turn_id": payload.get("turn_id"), "turn_index": turn_index, "steps": []}
                cur_step, tools_by_id, pending_tool_outputs = None, {}, {}
            elif pt == "user_message" and cur is not None:
                cur["user_input"] = payload.get("message")
            elif pt == "agent_message" and cur is not None:
                cur["final_output"] = payload.get("message")
            elif pt == "token_count" and cur is not None:
                info = payload.get("info") or {}
                cur["total_usage"] = _usage(info.get("total_token_usage"))
                if cur_step is not None:
                    cur_step["usage"] = _usage(info.get("last_token_usage"))
                finish_step()  # token_count delimits a model step
            elif pt == "task_complete":
                finish_turn(completed=True, aborted=False)
            elif pt == "turn_aborted":
                finish_turn(completed=True, aborted=True)
            elif pt.endswith("_end") and payload.get("call_id"):
                tc = tools_by_id.get(payload.get("call_id"))
                output = tool_output(payload)
                error = _tool_error(payload) if payload.get("status") in ("failed", "declined") else None
                if tc is not None:
                    if error:
                        tc["error"] = error
                    if not tc.get("output"):
                        tc["output"] = output
                else:
                    pending_tool_outputs[payload.get("call_id")] = {"output": output, "error": error}

    if cur is not None:
        finish_turn(completed=False, aborted=False)  # flush trailing in-progress turn
    return turns, session_meta


# ---------- collection-time reconstruction ----------

def find_rollouts(sessions_dir, since_mtime=0.0, before_mtime=None):
    paths = []
    for p in glob.glob(os.path.join(sessions_dir, "**", "rollout-*.jsonl"), recursive=True):
        try:
            mtime = os.path.getmtime(p)
            if mtime >= since_mtime and (before_mtime is None or mtime < before_mtime):
                paths.append(p)
        except OSError:
            continue
    return sorted(paths, key=lambda p: os.path.getmtime(p))


def reconstruct(
    sessions_dir,
    out_path,
    since_mtime=0.0,
    before_mtime=None,
    braintrust=False,
    run_id=None,
    source="codex",
    session_name=None,
):
    rollouts = find_rollouts(sessions_dir, since_mtime, before_mtime)
    sessions = []
    for rp in rollouts:
        try:
            turns, meta = parse_rollout(rp)
        except Exception as e:
            print(f"[codex_rollout] skip {rp}: {e}")
            continue
        if not turns:
            continue
        sessions.append({"rollout": rp, "session_meta": meta, "turns": turns})
        if braintrust:
            try:
                sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
                import bt_emit
                bt_emit.log_session(
                    turns,
                    meta,
                    run_id=run_id,
                    source=source,
                    name=session_name,
                )
            except Exception as e:
                print(f"[codex_rollout] braintrust push failed: {e}")
    payload = {
        "agent": "codex",
        "run_id": run_id,
        "source": source,
        "session_name": session_name,
        "n_sessions": len(sessions),
        "sessions_dir": sessions_dir,
        "since_mtime": since_mtime,
        "before_mtime": before_mtime,
        "sessions": sessions,
    }
    pathlib.Path(out_path).parent.mkdir(parents=True, exist_ok=True)
    with open(out_path, "w") as f:
        json.dump(payload, f, indent=2, default=str)
    print(f"[codex_rollout] {len(sessions)} session(s) -> {out_path}")
    return payload


def main():
    if len(sys.argv) >= 3 and sys.argv[1] == "--parse-only":
        turns, meta = parse_rollout(sys.argv[2])
        print(json.dumps({"session_meta": meta, "turns": turns}, indent=2, default=str))
        return
    ap = argparse.ArgumentParser(description="Reconstruct Codex agent traces from rollouts.")
    ap.add_argument("--sessions-dir",
                    default=os.path.join(os.environ.get("CODEX_HOME", os.path.expanduser("~/.codex")), "sessions"))
    ap.add_argument("--out", required=True)
    ap.add_argument("--since-mtime", type=float, default=0.0,
                    help="epoch seconds; only rollouts modified at/after are included")
    ap.add_argument("--before-mtime", type=float, default=None,
                    help="epoch seconds; only rollouts modified before this time are included")
    ap.add_argument("--braintrust", action="store_true")
    ap.add_argument("--run-id", default=None)
    ap.add_argument("--source", default=os.environ.get("TRACE_SOURCE", "codex"))
    ap.add_argument("--session-name", default=os.environ.get("TRACE_SESSION_NAME"))
    args = ap.parse_args()
    reconstruct(
        args.sessions_dir,
        args.out,
        args.since_mtime,
        args.before_mtime,
        args.braintrust,
        args.run_id,
        args.source,
        args.session_name,
    )


if __name__ == "__main__":
    main()

File: eval_harness/utils/telemetry/tracing/codex_rollout.ts
#!/usr/bin/env bun

/**
 * Reconstruct Codex rollout JSONL into the harness's normalized
 * turns -> steps -> tool_calls trace shape.
 */

import { mkdir, readdir, realpath, stat, writeFile } from "node:fs/promises";
import type { Dirent } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";

import {
  emitBraintrustSession,
  hasMeaningfulValue,
  parsePythonFloat,
  parsePythonJson,
  readJsonlRecords,
  stringifyPythonJson,
  type JsonRecord,
  type ToolCall,
  type TraceStep,
  type TraceTurn,
} from "./cc_transcript.ts";

type CodexReconstructionOptions = {
  sessionsDir: string;
  outPath: string;
  sinceMtime?: number;
  beforeMtime?: number;
  braintrust?: boolean;
  runId?: string | null;
  source?: string;
  sessionName?: string | null;
};

type PendingOutput = { output: unknown; error: unknown };

function optionalRecord(value: unknown): JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as JsonRecord)
    : {};
}

function recordOrPythonFallback(value: unknown, field: string): JsonRecord {
  if (!hasMeaningfulValue(value)) return {};
  if (typeof value === "object" && value !== null && !Array.isArray(value)) {
    return value as JsonRecord;
  }
  throw new TypeError(`${field} must be an object when present`);
}

function firstMeaningfulValue(...values: unknown[]): unknown {
  return values.find((value) => hasMeaningfulValue(value));
}

function extractText(content: unknown): string {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  const output: string[] = [];
  for (const part of content) {
    if (typeof part === "string") {
      output.push(part);
      continue;
    }
    const record = optionalRecord(part);
    const value = firstMeaningfulValue(record.text, record.output_text, record.input_text, "");
    output.push(typeof value === "string" ? value : "");
  }
  return output.join("");
}

function extractReasoning(payload: JsonRecord): string {
  if (hasMeaningfulValue(payload.content)) return extractText(payload.content);
  if (Array.isArray(payload.summary)) {
    return payload.summary.map((part) => extractText(part)).join(" ");
  }
  return "";
}

function isWrapper(text: string): boolean {
  const trimmed = text.trimStart();
  return trimmed.startsWith("<environment_context>") || trimmed.startsWith("<user_instructions>");
}

function maybeJson(value: unknown): unknown {
  if (typeof value === "object" && value !== null) return value;
  if (typeof value === "string") {
    try {
      return parsePythonJson(value);
    } catch {
      return value;
    }
  }
  return value;
}

function usage(value: unknown): JsonRecord {
  const source = optionalRecord(value);
  const result: JsonRecord = {};
  if (source.input_tokens !== null && source.input_tokens !== undefined) {
    result.prompt_tokens = source.input_tokens;
  }
  if (source.output_tokens !== null && source.output_tokens !== undefined) {
    result.completion_tokens = source.output_tokens;
  }
  if (source.total_tokens !== null && source.total_tokens !== undefined) {
    result.tokens = source.total_tokens;
  }
  if (source.cached_input_tokens !== null && source.cached_input_tokens !== undefined) {
    result.cache_read_input_tokens = source.cached_input_tokens;
  }
  return result;
}

function toolError(payload: JsonRecord): unknown {
  return firstMeaningfulValue(
    payload.error,
    payload.codex_error_info,
    payload.stderr,
    payload.aggregated_output,
    hasMeaningfulValue(payload.exit_code) ? `Exit code: ${String(payload.exit_code)}` : "failed",
  );
}

function toolOutput(payload: JsonRecord): unknown {
  const direct = firstMeaningfulValue(
    payload.aggregated_output,
    payload.stdout,
    payload.result,
    payload.output,
  );
  if (direct !== undefined) return direct;
  const fallback: JsonRecord = {};
  for (const key of ["status", "query", "action", "call_id"]) {
    if (payload[key] !== null && payload[key] !== undefined) fallback[key] = payload[key];
  }
  return fallback;
}

export async function parseRollout(path: string): Promise<[TraceTurn[], JsonRecord]> {
  let sessionMeta: JsonRecord = {};
  const turns: TraceTurn[] = [];
  let current: TraceTurn | undefined;
  let currentStep: TraceStep | undefined;
  let toolsById = new Map<unknown, ToolCall>();
  let pendingOutputs = new Map<unknown, PendingOutput>();
  let turnIndex = 0;

  const finishStep = (): void => {
    if (current !== undefined && currentStep !== undefined) {
      current.steps.push(currentStep);
      currentStep = undefined;
    }
  };

  const finishTurn = (completed: boolean, aborted: boolean): void => {
    if (current === undefined) return;
    finishStep();
    current.completed = completed;
    current.aborted = aborted;
    if (!hasMeaningfulValue(current.final_output)) {
      for (const step of [...current.steps].reverse()) {
        if (hasMeaningfulValue(step.text)) {
          current.final_output = step.text;
          break;
        }
      }
    }
    if (!hasMeaningfulValue(current.user_input)) {
      current.user_input = current.user_input_fallback ?? null;
    }
    turns.push(current);
    current = undefined;
    toolsById = new Map();
    pendingOutputs = new Map();
  };

  const attachPending = (toolCall: ToolCall): void => {
    const callId = toolCall.call_id;
    if (!hasMeaningfulValue(callId) || !pendingOutputs.has(callId)) return;
    const pending = pendingOutputs.get(callId);
    pendingOutputs.delete(callId);
    if (pending === undefined) return;
    if (hasMeaningfulValue(pending.error)) toolCall.error = pending.error;
    if (pending.output !== null && pending.output !== undefined) toolCall.output = pending.output;
  };

  for await (const record of readJsonlRecords(path)) {
    const type = record.type;
    if (
      type !== "session_meta" &&
      type !== "turn_context" &&
      type !== "response_item" &&
      type !== "event_msg"
    ) {
      continue;
    }
    const payload = recordOrPythonFallback(record.payload, "payload");

    if (type === "session_meta") {
      sessionMeta = payload;
      continue;
    }
    if (type === "turn_context" && current !== undefined) {
      current.model = firstMeaningfulValue(payload.model, current.model) ?? null;
      current.invocation_params = payload;
      continue;
    }
    if (type === "response_item") {
      if (current === undefined) continue;
      if (currentStep === undefined) {
        currentStep = { text: "", reasoning: "", tool_calls: [] };
      }
      const payloadType = payload.type;
      if (payloadType === "message") {
        const role = payload.role;
        const text = extractText(payload.content);
        if (role === "assistant") {
          currentStep.text += text;
        } else if ((role === "user" || role === "developer") && !isWrapper(text)) {
          if (!("user_input_fallback" in current)) current.user_input_fallback = text;
        }
      } else if (payloadType === "reasoning") {
        currentStep.reasoning += extractReasoning(payload);
      } else if (payloadType === "function_call" || payloadType === "custom_tool_call") {
        const argumentSource =
          payload.arguments !== null && payload.arguments !== undefined
            ? payload.arguments
            : payload.input;
        const toolCall: ToolCall = {
          call_id: payload.call_id ?? null,
          name: payload.name ?? null,
          args: maybeJson(argumentSource ?? null),
        };
        currentStep.tool_calls.push(toolCall);
        if (hasMeaningfulValue(toolCall.call_id)) {
          toolsById.set(toolCall.call_id, toolCall);
          attachPending(toolCall);
        }
      } else if (payloadType === "web_search_call") {
        const action = recordOrPythonFallback(payload.action, "action");
        const toolCall: ToolCall = {
          call_id: payload.id ?? null,
          name: "web_search",
          args: {
            query: action.query ?? null,
            queries: action.queries ?? null,
            action,
          },
        };
        currentStep.tool_calls.push(toolCall);
        if (hasMeaningfulValue(toolCall.call_id)) {
          toolsById.set(toolCall.call_id, toolCall);
          attachPending(toolCall);
        }
      } else if (payloadType === "function_call_output" || payloadType === "custom_tool_call_output") {
        const toolCall = toolsById.get(payload.call_id);
        if (toolCall !== undefined) toolCall.output = payload.output ?? null;
      }
      continue;
    }
    if (type !== "event_msg") continue;

    const payloadType = payload.type;
    if (payloadType === "task_started") {
      if (current !== undefined) finishTurn(false, false);
      turnIndex += 1;
      current = {
        turn_id: payload.turn_id ?? null,
        turn_index: turnIndex,
        steps: [],
        user_input: undefined,
      };
      currentStep = undefined;
      toolsById = new Map();
      pendingOutputs = new Map();
    } else if (payloadType === "user_message" && current !== undefined) {
      current.user_input = payload.message ?? null;
    } else if (payloadType === "agent_message" && current !== undefined) {
      current.final_output = payload.message ?? null;
    } else if (payloadType === "token_count" && current !== undefined) {
      const info = recordOrPythonFallback(payload.info, "info");
      current.total_usage = usage(info.total_token_usage);
      if (currentStep !== undefined) currentStep.usage = usage(info.last_token_usage);
      finishStep();
    } else if (payloadType === "task_complete") {
      finishTurn(true, false);
    } else if (payloadType === "turn_aborted") {
      finishTurn(true, true);
    } else if (
      typeof payloadType === "string" &&
      payloadType.endsWith("_end") &&
      hasMeaningfulValue(payload.call_id)
    ) {
      const toolCall = toolsById.get(payload.call_id);
      const output = toolOutput(payload);
      const error =
        payload.status === "failed" || payload.status === "declined"
          ? toolError(payload)
          : null;
      if (toolCall !== undefined) {
        if (hasMeaningfulValue(error)) toolCall.error = error;
        if (!hasMeaningfulValue(toolCall.output)) toolCall.output = output;
      } else {
        pendingOutputs.set(payload.call_id, { output, error });
      }
    }
  }

  if (current !== undefined) finishTurn(false, false);
  return [turns, sessionMeta];
}

async function walkRollouts(
  directory: string,
  ancestorDirectories: ReadonlySet<string> = new Set(),
): Promise<string[]> {
  const paths: string[] = [];
  let resolvedDirectory: string;
  let entries: Dirent<string>[];
  try {
    resolvedDirectory = await realpath(directory);
    if (ancestorDirectories.has(resolvedDirectory)) return [];
    entries = await readdir(directory, { withFileTypes: true });
  } catch {
    return [];
  }
  const nextAncestors = new Set(ancestorDirectories);
  nextAncestors.add(resolvedDirectory);
  for (const entry of entries) {
    const path = join(directory, entry.name);
    try {
      const target = await stat(path);
      if (target.isDirectory()) paths.push(...(await walkRollouts(path, nextAncestors)));
      else if (
        target.isFile() &&
        entry.name.startsWith("rollout-") &&
        entry.name.endsWith(".jsonl")
      ) {
        paths.push(path);
      }
    } catch {
      // A broken/inaccessible entry does not discard readable siblings.
    }
  }
  return paths;
}

export async function findRollouts(
  sessionsDir: string,
  sinceMtime = 0,
  beforeMtime?: number,
): Promise<string[]> {
  let candidates: string[];
  try {
    candidates = await walkRollouts(sessionsDir);
  } catch {
    return [];
  }
  const selected: Array<{ path: string; mtime: number }> = [];
  for (const path of candidates) {
    try {
      const mtime = (await stat(path)).mtimeMs / 1_000;
      if (mtime >= sinceMtime && (beforeMtime === undefined || mtime < beforeMtime)) {
        selected.push({ path, mtime });
      }
    } catch {
      // Ignore files that disappear during collection.
    }
  }
  selected.sort((left, right) => left.mtime - right.mtime);
  return selected.map(({ path }) => path);
}

export async function reconstructCodexRollouts(
  options: CodexReconstructionOptions,
): Promise<JsonRecord & { sessions: JsonRecord[] }> {
  const sinceMtime = options.sinceMtime ?? 0;
  const source = options.source ?? "codex";
  const runId = options.runId ?? null;
  const sessionName = options.sessionName ?? null;
  const rollouts = await findRollouts(options.sessionsDir, sinceMtime, options.beforeMtime);
  const sessions: JsonRecord[] = [];
  for (const rollout of rollouts) {
    try {
      const [turns, metadata] = await parseRollout(rollout);
      if (turns.length === 0) continue;
      sessions.push({ rollout, session_meta: metadata, turns });
      if (options.braintrust) {
        await emitBraintrustSession(
          turns,
          metadata,
          runId,
          source,
          sessionName,
          "codex_rollout",
        );
      }
    } catch (error) {
      console.log(`[codex_rollout] skip ${rollout}: ${String(error)}`);
    }
  }
  const payload = {
    agent: "codex",
    run_id: runId,
    source,
    session_name: sessionName,
    n_sessions: sessions.length,
    sessions_dir: options.sessionsDir,
    since_mtime: sinceMtime,
    before_mtime: options.beforeMtime ?? null,
    sessions,
  };
  await mkdir(dirname(options.outPath), { recursive: true });
  await writeFile(options.outPath, stringifyPythonJson(payload, 2));
  console.log(`[codex_rollout] ${sessions.length} session(s) -> ${options.outPath}`);
  return payload;
}

type ParsedArguments = {
  sessionsDir: string;
  outPath?: string;
  sinceMtime: number;
  beforeMtime?: number;
  braintrust: boolean;
  runId: string | null;
  source: string;
  sessionName: string | null;
};

const CODEX_HELP = `usage: codex_rollout.py [-h] [--sessions-dir SESSIONS_DIR] --out OUT
                        [--since-mtime SINCE_MTIME]
                        [--before-mtime BEFORE_MTIME] [--braintrust]
                        [--run-id RUN_ID] [--source SOURCE]
                        [--session-name SESSION_NAME]

Reconstruct Codex agent traces from rollouts.

options:
  -h, --help            show this help message and exit
  --sessions-dir SESSIONS_DIR
  --out OUT
  --since-mtime SINCE_MTIME
                        epoch seconds; only rollouts modified at/after are
                        included
  --before-mtime BEFORE_MTIME
                        epoch seconds; only rollouts modified before this time
                        are included
  --braintrust
  --run-id RUN_ID
  --source SOURCE
  --session-name SESSION_NAME`;

const CODEX_OPTIONS = [
  "--help",
  "--sessions-dir",
  "--out",
  "--since-mtime",
  "--before-mtime",
  "--braintrust",
  "--run-id",
  "--source",
  "--session-name",
] as const;

function resolveOption(name: string): (typeof CODEX_OPTIONS)[number] {
  const candidates = CODEX_OPTIONS.filter(
    (option) => option === name || option.startsWith(name),
  );
  if (candidates.length !== 1) {
    throw new Error(
      candidates.length === 0
        ? `unrecognized argument: ${name}`
        : `ambiguous option: ${name} could match ${candidates.join(", ")}`,
    );
  }
  return candidates[0] as (typeof CODEX_OPTIONS)[number];
}

function parseArguments(args: string[]): ParsedArguments | "help" {
  const codexHome = process.env.CODEX_HOME ?? join(homedir(), ".codex");
  const parsed: ParsedArguments = {
    sessionsDir: join(codexHome, "sessions"),
    sinceMtime: 0,
    braintrust: false,
    runId: null,
    source: process.env.TRACE_SOURCE ?? "codex",
    sessionName: process.env.TRACE_SESSION_NAME ?? null,
  };
  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index];
    if (argument === "-h") return "help";
    if (argument === undefined || !argument.startsWith("--")) {
      throw new Error(`unrecognized argument: ${argument ?? ""}`);
    }
    const equalsIndex = argument.indexOf("=");
    const optionText = equalsIndex === -1 ? argument : argument.slice(0, equalsIndex);
    const attachedValue = equalsIndex === -1 ? undefined : argument.slice(equalsIndex + 1);
    const option = resolveOption(optionText);
    if (option === "--help") return "help";
    if (option === "--braintrust") {
      if (attachedValue !== undefined) throw new Error("argument --braintrust: ignored explicit argument");
      parsed.braintrust = true;
      continue;
    }
    const value = attachedValue ?? args[index + 1];
    if (value === undefined) {
      throw new Error(`argument ${option}: expected one argument`);
    }
    if (option === "--sessions-dir") parsed.sessionsDir = value;
    else if (option === "--out") parsed.outPath = value;
    else if (option === "--since-mtime" || option === "--before-mtime") {
      const numeric = parsePythonFloat(value);
      if (numeric === undefined) throw new Error(`argument ${option}: invalid float value: '${value}'`);
      if (option === "--since-mtime") parsed.sinceMtime = numeric;
      else parsed.beforeMtime = numeric;
    } else if (option === "--run-id") parsed.runId = value;
    else if (option === "--source") parsed.source = value;
    else if (option === "--session-name") parsed.sessionName = value;
    if (attachedValue === undefined) index += 1;
  }
  return parsed;
}

export async function main(args: string[] = process.argv.slice(2)): Promise<number> {
  if (args[0] === "--parse-only" && args[1] !== undefined) {
    const [turns, metadata] = await parseRollout(args[1]);
    console.log(stringifyPythonJson({ session_meta: metadata, turns }, 2));
    return 0;
  }
  try {
    const parsed = parseArguments(args);
    if (parsed === "help") {
      console.log(CODEX_HELP);
      return 0;
    }
    if (parsed.outPath === undefined) throw new Error("the following arguments are required: --out");
    await reconstructCodexRollouts(parsed as CodexReconstructionOptions & { outPath: string });
    return 0;
  } catch (error) {
    console.error(`codex_rollout.ts: error: ${error instanceof Error ? error.message : String(error)}`);
    return 2;
  }
}

if (import.meta.main) {
  process.exit(await main());
}
